src/storage-controller/src/lib.rs

```diff
@@ -3199,11 +3237,58 @@ where
                 // Re-pack all rows
                 let mut packer = row_buf.packer();
                 packer.extend(unpacked_row.into_iter());
-                (row_buf.clone(), -1)
+
+                let update = Update {
+                    row: row_buf.clone(),
+                    timestamp: expected_upper.clone(),
+                    diff: -1,
+                };
+
+                update
             })
             .collect();
 
-        self.append_to_managed_collection(id, updates).await;
+        let command = (id, updates, expected_upper, new_upper);
+
+        let res = self
+            .persist_monotonic_worker
+            .compare_and_append(vec![command])
+            .await
+            .expect("command must succeed");
+
+        match res {
+            Ok(_) => {
+                // All good, yay!
+            }
+            Err(storage_err) => {
+                match storage_err {
+                    StorageError::InvalidUppers(failed_ids) => {
+                        assert_eq!(
+                            failed_ids.len(),
+                            1,
+                            "received errors for more than one collection"
+                        );
+                        assert_eq!(
+                            failed_ids[0].id, id,
+                            "received errors for a different collection"
+                        );
+
+                        // This is fine, it just means the upper moved because
+                        // of continual upper advancement or because seomeone
+                        // already appended some more retractions/updates.
+                        //
+                        // NOTE: We might want to attempt these partial
+                        // retractions on an interval, instead of only when
+                        // starting up!
+                        info!(%id, current_upper = ?failed_ids[0].current_upper, "failed to append partial truncation");
+                    }
+                    // Uh-oh, something else went wrong!
+                    other => {
+                        panic!("Unhandled error while appending to managed collection {id:?}: {other:?}")
+                    }
+                }
+            }
+        }
```

Comment:
Does not seem to be covered by tests according to [coverage/428](https://buildkite.com/materialize/coverage/builds/428).

Reply:
Yeah, this also can't happen today because we can't have concurrent `envd` processes which are attempting to write. We might see this when we have real 0dt upgrades, where we temporarily have two `envd` processes alive.

So it's very much future-looking code.